File: dashboard/src/types/machine.ts
/*
    ID           string   `json:"id"`           // The ID of the machine
	OS           string   `json:"os"`           //The operating system of the machine
	Hostname     string   `json:"hostname"`     //Hostname of the machine
	IPAddresses  []string `json:"ip_addresses"` //A list of ip addreses of the machine on all network interfaces
	NumberAgents int64    `json:"numberAgents"` //The number of agents the machine has on itself
}
*/
type Machine = {
    id: string,
    os: string,
    hostname: string,
    ipAddresses: string[],
    numberAgents: number
}

type MachineResponse = {
    machines: Machine[]
}

interface MachineProps {
    machine: Machine
}

interface OSProps {
    os: string
}

type MachineRegisterResponse = {
    uuid: string
}

// type MachinesStatisticsResponse struct {
// 	TotalMachines   int64 `json:"totalMachines"`
// 	TotalInterfaces int64 `json:"totalInterfaces"`
// }

type MachinesStatisticsResponse = {
    totalMachines: number,
    totalInterfaces: number
}

File: api/data/machine.go
package data

import (
	"encoding/json"
	"io"
)

const (
	UNKOWN  int64 = 0
	WINDOWS int64 = 1
	LINUX   int64 = 2
)

// Holds data collected by the proxy about the machine
type MachineInformation struct {
	OS          string   `json:"os"`          //The operating system of the machine
	Hostname    string   `json:"hostname"`    //Hostname of the machine
	IPAddresses []string `json:"ipAddresses"` //A list of ip addreses of the machine on all network interfaces
}

func (mach *MachineInformation) ToJSON(w io.Writer) error {
	e := json.NewEncoder(w)
	return e.Encode(mach)
}

func (mach *MachineInformation) FromJSON(r io.Reader) error {
	d := json.NewDecoder(r)
	return d.Decode(mach)
}

type MachineDatabase struct {
	ID           string   `json:"id"`           // The ID of the machine
	OS           string   `json:"os"`           //The operating system of the machine
	Hostname     string   `json:"hostname"`     //Hostname of the machine
	IPAddresses  []string `json:"ipAddresses"`  //A list of ip addreses of the machine on all network interfaces
	NumberAgents int64    `json:"numberAgents"` //The number of agents the machine has on itself
}

func (mach *MachineDatabase) ToJSON(w io.Writer) error {
	e := json.NewEncoder(w)
	return e.Encode(mach)
}

func (mach *MachineDatabase) FromJSON(r io.Reader) error {
	d := json.NewDecoder(r)
	return d.Decode(mach)
}
